src/parser/parser_driver.cpp

```diff
@@ -433,51 +435,96 @@ void ParserDriver::defineGrammar()
+  	_lastRuleLocation = currentFileContext()->getLocation();
+	  _lastRuleTokenStream = currentFileContext()->getTokenStream();
+		return {};
+	};
+
+  auto const common_rule_init = [&](auto&& args) -> Value {
+		const std::string& name_text = args[3].getTokenIt()->getString();
+		if (ruleExists(name_text))
+		  error_handle(args[3].getTokenIt()->getLocation(), "Redefinition of rule '" + args[3].getTokenIt()->getString() + "'");
+		args[3].getTokenIt()->setType(TokenType::RULE_NAME);
+		args[3].getTokenIt()->setValue(std::make_shared<ValueSymbol>(name_text, Expression::Type::Bool));
+		return {};
+	};
+
+  if (_features & Features::AvastOnly)
+    _parser.rule("rule") // {}
+		  .production(
+			  "rule_mods", "RULE", common_last_rule, "ID", common_rule_init, "tags", "LCB", "metas", "strings", "variables", "condition" , "RCB", [&](auto&& args) -> Value {
+	        std::optional<TokenIt> mod_private = {};
+    		  std::optional<TokenIt> mod_global = {};
+		      const std::vector<TokenIt> mods = std::move(args[0].getMultipleTokenIt());
+      		for (const auto &token: mods)
+		      {
+      			if (token->getType() == TokenType::GLOBAL)
+			      {
+      				if (mod_global.has_value())
+			      		error_handle(token->getLocation(), "Duplicated global rule modifier");
+      				mod_global = token;
+			      }
+      			else if (token->getType() == TokenType::PRIVATE)
+			      {
+      				if (mod_private.has_value())
+			      		error_handle(token->getLocation(), "Duplicated private rule modifier");
+      				mod_private = token;
+			      }
+      		}
+      		TokenIt name = args[3].getTokenIt();
+		      const std::vector<TokenIt> tags = std::move(args[5].getMultipleTokenIt());
+      		args[6].getTokenIt()->setType(TokenType::RULE_BEGIN);
+		      std::vector<Meta> metas = std::move(args[7].getMetas());
+		      std::shared_ptr<Rule::StringsTrie> strings = std::move(args[8].getStringsTrie());
+		      std::vector<Variable> variables = std::move(args[9].getVariables());
+		      Expression::Ptr condition = std::move(args[10].getExpression());
+		      args[11].getTokenIt()->setType(TokenType::RULE_END);
+
+		      for(auto iter = variables.begin(); iter != variables.end(); iter++) {
+			      removeLocalSymbol(iter->getKey());
+		      }
+
+          addRule(Rule(_lastRuleTokenStream, name, std::move(mod_private), std::move(mod_global),
+			      std::move(metas), std::move(strings), std::move(variables), std::move(condition), std::move(tags)));
+		      return {};
+      })
+      ;
+  else
+	  _parser.rule("rule") // {}
+		  .production(
+			  "rule_mods", "RULE", common_last_rule, "ID", common_rule_init, "tags", "LCB", "metas", "strings", "condition" , "RCB", [&](auto&& args) -> Value {
+		      std::optional<TokenIt> mod_private = {};
+  		    std::optional<TokenIt> mod_global = {};
+	  	    const std::vector<TokenIt> mods = std::move(args[0].getMultipleTokenIt());
+		      for (const auto &token: mods)
```

Comment:
This is kinda better in a sense that at least it's in condition for Avast features but can we at least simplify this solution so that we don't repeat the same code? If you are not able to do that via `std::bind` and lambdas then at least please introduce some free functions that would do these bits atomically and we can reuse them? Imagine that there's a bug in this for loop for some reason. You'd have to change it on multiple places and that is heavily error prone. Thank you.